tests/gold_tests/redirect/redirect-handle-abnormal-address.test.py

```diff
@@ -0,0 +1,165 @@
+        })
+        tr.Processes.Default.StartBefore(trafficservers[scenario])
+    else:
+        tr.StillRunningAfter = trafficservers[scenario]
+
+    testDomain = 'testdomain{0}.test'.format(normRedirectTarget)
+    # The micro DNS server can't tell us whether it has a record of the domain already, so we use a dictionary to avoid duplicates.
+    # We remove any surrounding brackets that are common to IPv6 addresses.
+    dnsRecords[testDomain] = [redirectTarget.strip('[]')]
+
+    # A GET request parameterized on the scenario and on the target.
+    request_header = {
+            'headers': ('GET /redirect?scenario={0}&target={1} HTTP/1.1\r\n'
+                        'Host: *\r\n\r\n').\
+                                format(scenario.name, normRedirectTarget),
+            'timestamp': ArbitraryTimestamp,
+            'body': ''}
+    # Returns a redirect to the test domain for the given target & the port number for the TS of the given scenario.
+    response_header = {
+            'headers': ('HTTP/1.1 307 Temporary Redirect\r\n'
+                        'Location: http://{0}:{1}/\r\n'
+                        'Connection: close\r\n\r\n').\
+                                format(testDomain, origin.Variables.Port),
+            'timestamp': ArbitraryTimestamp,
+            'body': ''}
+    origin.addResponse('sessionfile.log', request_header, response_header)
+
+    # Generate the request data file.
+    with open(os.path.join(data_path, tr.Name), 'w') as f:
+        f.write(('GET /redirect?scenario={0}&target={1} HTTP/1.1\r\n'
+                 'Host: iwillredirect.test:{2}\r\n\r\n').\
+                         format(scenario.name, normRedirectTarget, origin.Variables.Port))
+    # Set the command with the appropriate URL.
+    tr.Processes.Default.Command = 'python tcp_client.py 127.0.0.1 {0} {1} | head -n 1'.\
+            format(trafficservers[scenario].Variables.port, os.path.join(data_dirname, tr.Name))
+    # Generate and set the 'gold file' to check stdout
+    goldFilePath = os.path.join(data_path, '{0}.gold'.format(tr.Name))
+    with open(goldFilePath, 'w') as f:
+        f.write(scenario.value['expectedStatusLine'])
+    tr.Processes.Default.Streams.stdout = goldFilePath
+
+host = socket.gethostname()
+ipv4addrs = [ip for \
+        (family,_,_,_,(ip,*_)) in \
+        socket.getaddrinfo(host,port=None) if \
+        socket.AF_INET == family]
+ipv6addrs = ["[{0}]".format(ip.split('%')[0]) for \
+        (family,_,_,_,(ip,*_)) in \
+        socket.getaddrinfo(host,port=None) if \
+        socket.AF_INET6 == family and 'fe80' != ip[0:4]] # Skip link-local addresses.
+
+for t in ipv4addrs + ipv6addrs:
+    for s in ScenarioE:
+        makeTestCase(redirectTarget=t, scenario=s)
+
+for t in (
+        '127.1.2.3', '[::1]',      # loopback
+        '169.254.0.1', '[fe80::]', # link-local
+        '224.1.2.3', '[ff42::]',   # multicast
+        '10.0.0.1', '[fc00::1]',   # private
```

Comment:
NOTE: These are not exhaustive. I am not testing whether our detection of all the possible values is correct, only that we behave correctly when configured to detect such classes.